### ISSUE TITLE
Consider Blob.fromStream (returns a Promise<Blob>)

### ISSUE THREAD

**Issue by domenic** [w3c/FileAPI#140]
I'm filing this as a tracking issue, although I don't think it's particularly urgent. But maybe web developers have run into this more, in which case hearing from them would be great.

Anyway, right now, to convert a `ReadableStream` to a `Blob`, you have to do

```
const blob = await new Response(stream).blob();
```

which is pretty silly. Just as `blob.stream()` was added to avoid the silly `new Request(blob).body` pattern, perhaps we should consider adding a promise-returning `Blob.fromStream()` so you could do

```js
const blob = await Blob.fromStream(stream);
```

This is a bigger savings if we add an `options` parameter:

```js
const blob1 = new Blob([await new Response(stream).blob()], options);
const blob2 = Blob.fromStream(stream, options);
```

Points to consider:

- We could instead make this `stream.toBlob()`, but I feel the layering of keeping streams ignorant of blobs is a bit cleaner.
- We may not want to introduce this because it's better if people avoid using blobs? (Or is it just blob URLs that we dislike?)
- If someone wanted a `File`, we could either add `File.fromStream(stream, fileName, options)` or we could have people do `new File([Blob.fromStream(stream)], fileName, options)`. Probably `File.fromStream()` is nicer.

**Comment by annevk** [comment]
We only dislike `blob:` URLs (and have a solution, but nobody with the time to refactor all relevant entry points) as far as I know.

**Comment by mkruisselbrink** [comment]
Yep, blob URLs are bad, both blobs in general are fine. I'd be supportive of adding a Blob.fromStream method.

**Comment by guest271314** [comment]
AFAICT 

`const blob = await new Response(stream).blob();` 

throws a `TypeError` 

`Promise {<rejected>: TypeError: Failed to fetch}`

The concept appears to be similar what is possible with Native File System, e.g., 

```
const fileFromStream = async(stream, fileName) =>  {
  const dir = await FileSystemDirectoryHandle.getSystemDirectory({ type: 'sandbox' });
  await stream
    .pipeTo(await (await dir.getFile(fileName, { create: true }))
    .createWritable(), { preventCancel: true });
  const file = (await dir.getFile(fileName)).getFile();
  // cleanup potential lingering temporary file (until page refreshed) in sandbox
  await dir.removeEntry(fileName);
  // sanity check
  for await(let entry of dir.getEntries()) {
    console.log(await entry);
    if (entry.isDirectory) await entry.removeRecursively();
    else await dir.removeEntry(fileName);
  }
  return file;
}

fileFromStream(new ReadableStream({start(c) {c.enqueue('1'); c.close()}}), 'file')
.then(console.log, console.error);
```

**Comment by guest271314** [comment]
@mkruisselbrink 

> Yep, blob URLs are bad

Can you explain why? 

What is the available substitute?

**Comment by jimmywarting** [comment]
In `node-fetch` / `fetch-blob` package I have been figuring about such method myself.
I found that NodeJS new stream consumer `(node:stream/consumers).blob(iterable)` to be somewhat useful but one thing i did not like about it was that I had to do `new Blob([part], { type: contentType })` afterwards, b/c it currently lacks options for it...

One thing i found useful about this proposal is that we could then have a way of creating quite large blobs backed up by the file system if this is needed. (could be useful for eg FileSaver.js)

---

if we could hint about a known size then there wouldn't even be a reason for a person to have to wait for the concatination to finish either, you could return a blob directly.
```js
const blob = Blob.fromStream(stream, { size: 1024 })
```

I think that the ability to generate your own blob backed with whatever method you may prefer would be a neat feature.

Another pretty useful thing i built in `fetch-blob` was the ability to create a File/Blob-like item backed up by the filesystem that could be handled by the Blob class itself, (now undici have adapter theirs FormData impl to be able to accept any 3th party BlobLike items also - same as node-fetch's FormData implementation)

so instead of having `blob = await blobFromPath('./readme.md')` i could maybe have something like:

```js
var blob = Blob.from({
  size: 1024,

  type: 'text/plain',

  slice(start, end) {
    // return a new blob with a different size/offset
    return Blob.from(...)
  },

  /**
   * return your own readable stream method, that could
   * technically be called multiple times over and over.
   * 
   * @return {ReadableStream<Uint8Array>}
   */
  stream() {
    // create your own readable stream method that returns the data
    const ts = new TransformStream({...})
    fetch(url).then(res => res.body.pipeTo(ts.writable))
    return ts.readable
  }
})
```

this is esentially what i have built in [fetch-blob](https://github.com/node-fetch/fetch-blob/blob/main/from.js#L59)
I have also made a [Blob/File like class out of a http request](https://github.com/transcend-io/conflux/issues/31) that supports byte ranges in order to parse zip files and download only the parts of the zip and unpack individual files from the entire thing.

if they could somehow be representable by a native Blob class then you could do so much more with them. like creating a Blob URL \w `URL.createObjectURL` and play it in a video element. but this can't work cuz they are only blob like items and you can't create other native blobs with this made up blob parts `new Blob( [ 'str', blobLikeItem ], { type: 'text/html' })` which is a bit sad

**Comment by jimmywarting** [comment]
> This does not allow creating a blob that represents an "in progress" stream, which I've heard people have wanted in order to have self-referential blobs. That would require new concepts and infrastructure. This proposal is basically just exposing existing infrastructure in a more straightforward way.

This is something i badly want to have right now. I feel like `Blob.from({...})` ☝️ could solve this problem and so much more. Either that or the option to accept other blob-like objects. aka: `new File([ blobLike ], fileName, options)`

**Comment by dmurph** [comment]
TPAC 2024: General consensus with Chromium & FF that this seems nice to have, but not urgent. No strong opinion from WebKit. Likely would impl in C++. Easy polyfill.
